chemprop/train/evaluate.py

```diff
@@ -31,68 +35,89 @@ def evaluate_predictions(preds: List[List[float]],
-                    nan = True
-                    info('Warning: Found a task with targets all 0s or all 1s')
-                if all(pred == 0 for pred in valid_preds[i]) or all(pred == 1 for pred in valid_preds[i]):
-                    nan = True
-                    info('Warning: Found a task with predictions all 0s or all 1s')
-
-                if nan:
-                    for metric in metrics:
-                        results[metric].append(float('nan'))
+            if dataset_type == "classification":
+                return_nan = False
+                if (task_targets == 0).all() or (task_targets == 1).all():
+                    return_nan = True
+                    info("Warning: Found a task with targets all 0s or all 1s")
+                if (task_preds == 0).all() or (task_preds == 1).all():
+                    return_nan = True
+                    info("Warning: Found a task with predictions all 0s or all 1s")
+
+                if return_nan:
+                    for metric_label in metrics:
+                        results[metric_label].append(float("nan"))
                     continue
 
-            if len(valid_targets[i]) == 0:
+            if len(task_targets) == 0:
                 continue
 
-            for metric, metric_func in metric_to_func.items():
-                if dataset_type == 'multiclass' and metric == 'cross_entropy':
-                    results[metric].append(metric_func(valid_targets[i], valid_preds[i],
-                                                    labels=list(range(len(valid_preds[i][0])))))
-                elif metric in ['bounded_rmse', 'bounded_mse', 'bounded_mae']:
-                    results[metric].append(metric_func(valid_targets[i], valid_preds[i], gt_targets[i], lt_targets[i]))
+            for metric_label, metric_func in metric_to_func.items():
+                if dataset_type == "multiclass" and metric_label == "cross_entropy":
+                    results[metric_label].append(
+                        metric_func(
+                            task_targets,
+                            task_preds,
+                            labels=list(range(task_preds.shape[1])),
+                        )
+                    )
+                elif metric_label in ["bounded_rmse", "bounded_mse", "bounded_mae"]:
+                    results[metric_label].append(
+                        metric_func(
+                            task_targets,
+                            task_preds,
+                            task_gt_targets,
+                            task_lt_targets,
+                        )
+                    )
                 else:
-                    results[metric].append(metric_func(valid_targets[i], valid_preds[i]))
+                    results[metric_label].append(
+                        metric_func(task_targets, task_preds)
+                    )
 
     results = dict(results)
```

Comment:
```suggestion
    return results
```
`results` is already of type `dict` (a subclass, specifically)